File: src/main/java/com/checkers/board/King.java
package com.checkers.board;

import com.checkers.exceptions.CannotEatException;
import com.checkers.exceptions.CannotMoveException;
import com.checkers.exceptions.MustEatException;
import com.checkers.utils.BoardUtils;
import com.checkers.utils.Color;

import java.util.Map;
import java.util.Optional;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class King extends Piece {

    public King(Color color) {
        super(color);
    }

    @Override
    public void analyzeAbilityOfMove() {
        var row = cell.getRow();
        var column = cell.getColumn();
        canMove = IntStream.iterate(1, i -> i + 1)
                .limit(7)
                .boxed()
                .flatMap(i -> Stream.of(Map.entry(i, i), Map.entry(i, -i), Map.entry(-i, -i), Map.entry(-i, i)))
                .filter(entry -> BoardUtils.isCoordinatesExists(row + entry.getKey(), column + entry.getValue()))
                .anyMatch(entry -> isAbleToMoveTo(cell.getNear(entry.getKey(), entry.getValue())));
    }

    @Override
    public void analyzeAbilityOfEat() {
        var row = cell.getRow();
        var column = cell.getColumn();
        canEat = IntStream.iterate(2, i -> i + 1)
                .limit(6)
                .boxed()
                .flatMap(i -> Stream.of(Map.entry(i, i), Map.entry(i, -i), Map.entry(-i, -i), Map.entry(-i, i)))
                .filter(entry -> BoardUtils.isCoordinatesExists(row + entry.getKey(), column + entry.getValue()))
                .anyMatch(entry -> isAbleToEatTo(cell.getNear(entry.getKey(), entry.getValue())));
    }

    @Override
    public boolean isAbleToMoveTo(Cell destinationCell) {
        if (cell.diff(destinationCell) == -1) {
            return false;
        }

        if (!destinationCell.isEmpty()) {
            return false;
        }

        if (isOnOtherDiagonal(destinationCell)) {
            return false;
        }

        var row = cell.getRow();
        var column = cell.getColumn();
        var destinationRow = destinationCell.getRow();
        var destinationColumn = destinationCell.getColumn();
        var deltaColumn = destinationColumn - column;
        var deltaRow = destinationRow - row;

        var deltaColumnSign = (int) Math.signum(deltaColumn);
        var deltaRowSign = (int) Math.signum(deltaRow);

        var board = cell.getBoard();
        for (var i = 1; ; i++) {
            var currentRow = row + deltaRowSign * i;
            var currentColumn = column + deltaColumnSign * i;

            if (currentRow == destinationRow && currentColumn == destinationColumn) {
                return true;
            }

            var currentCell = board.getCell(currentRow, currentColumn);
            if (!currentCell.isEmpty()) {
                return false;
            }
        }
    }

    @Override
    public boolean isAbleToEatTo(Cell destinationCell) {
        if (!destinationCell.isEmpty()) {
            return false;
        }

		if (isOnOtherDiagonal(destinationCell)) {
			return false;
		}

        if (cell.diff(destinationCell) < 2) {
            return false;
        }

        var row = cell.getRow();
        var column = cell.getColumn();
        var destinationRow = destinationCell.getRow();
        var destinationColumn = destinationCell.getColumn();
        var deltaColumn = destinationColumn - column;
        var deltaRow = destinationRow - row;

        var deltaColumnSign = (int) Math.signum(deltaColumn);
        var deltaRowSign = (int) Math.signum(deltaRow);

        var enemyPieceCount = 0;
        var board = cell.getBoard();
        for (var i = 1; ; i++) {
            var currentRow = row + deltaRowSign * i;
            var currentColumn = column + deltaColumnSign * i;

            if (currentColumn == destinationColumn && currentRow == destinationRow) {
                return enemyPieceCount == 1;
            }

            var currentCell = board.getCell(currentRow, currentColumn);
            if (currentCell.isEmpty()) {
                continue;
            }

            var piece = currentCell.getPiece();
            if (piece.getColor() == color) {
                return false;
            }

            enemyPieceCount++;
        }
    }

    private boolean isOnOtherDiagonal(Cell destinationCell) {
        var deltaColumn = destinationCell.getColumn() - cell.getColumn();
        var deltaRow = destinationCell.getRow() - cell.getRow();
        var k = (float) deltaColumn / deltaRow;
        return Math.abs(k) != 1;
    }

    @Override
    public void move(Cell destinationCell) {
        if (canEat) {
            throw new MustEatException();
        }

        if (!canMove) {
            throw new CannotMoveException(cell, destinationCell);
        }

        if (!isAbleToMoveTo(destinationCell)) {
            throw new CannotMoveException(cell, destinationCell);
        }

        destinationCell.setPiece(this);
        cell.setPiece(null);
    }

    @Override
    public void eat(Cell destinationCell) {
        if (!canEat) {
            throw new CannotEatException(cell, destinationCell);
        }

        if (!isAbleToEatTo(destinationCell)) {
            throw new CannotEatException(cell, destinationCell);
        }

        var sacrificedPiece = getSacrificedPiece(destinationCell)
                .orElseThrow(() -> new CannotEatException(cell, destinationCell));
        var sacrificedPieceCell = sacrificedPiece.getCell();
        sacrificedPieceCell.setPiece(null);
        destinationCell.setPiece(this);
        cell.setPiece(null);
    }

    private Optional<Piece> getSacrificedPiece(Cell destinationCell) {
        var destinationRow = destinationCell.getRow();
        var destinationColumn = destinationCell.getColumn();
        var sourceRow = cell.getRow();
        var sourceColumn = cell.getColumn();

        var signRow = (int) Math.signum(destinationRow - sourceRow);
        var signCol = (int) Math.signum(destinationColumn - sourceColumn);

        var board = cell.getBoard();
        for (var i = 1; ; i++) {
            var currentRow = sourceRow + signRow * i;
            var currentColumn = sourceColumn + signCol * i;

            if (currentRow == destinationRow && currentColumn == destinationColumn) {
                return Optional.empty();
            }

            var currentCell = board.getCell(currentRow, currentColumn);
            if (currentCell.isEmpty()) {
                continue;
            }

            var piece = currentCell.getPiece();
            if (piece.getColor() == color) {
                return Optional.empty();
            } else {
                return Optional.of(piece);
            }
        }
    }

    @Override
    public String toString() {
        if (color == Color.BLACK) {
            return "&";
        } else {
            return "#";
        }
    }

    @Override
    public boolean isMan() {
        return false;
    }

    @Override
    public boolean isKing() {
        return true;
    }
}

File: src/main/kotlin/com/checkers/board/King.kt
package com.checkers.board

import com.checkers.exceptions.CannotEatException
import com.checkers.exceptions.CannotMoveException
import com.checkers.exceptions.MustEatException
import com.checkers.utils.BoardUtils
import com.checkers.utils.Color
import kotlin.math.abs
import kotlin.math.sign

class King(color: Color) : Piece(color) {

    override fun analyzeAbilityOfMove() {
        canMove = generateSequence(1) { it + 1 }
            .take(7)
            .flatMap { sequenceOf(it to it, it to -it, -it to -it, -it to it) }
            .filter { BoardUtils.isCoordinatesExists(cell.row + it.first, cell.column + it.second) }
            .any { isAbleToMoveTo(cell.getNear(it.first, it.second)) }
    }

    override fun move(destinationCell: Cell) {
        if (canEat) {
            throw MustEatException()
        }

        if (!canMove || !isAbleToMoveTo(destinationCell)) {
            throw CannotMoveException(cell, destinationCell)
        }

        cell.piece = null
        destinationCell.piece = this
    }

    override fun isAbleToMoveTo(destinationCell: Cell): Boolean {
        if (cell.diff(destinationCell) == -1) {
            return false
        }

        if (destinationCell.piece != null) {
            return false
        }

        if (isOnOtherDiagonal(destinationCell)) {
            return false
        }

        val deltaColumn = destinationCell.column - cell.column
        val deltaRow = destinationCell.row - cell.row

        val deltaColumnSign = sign(deltaColumn.toDouble()).toInt()
        val deltaRowSign = sign(deltaRow.toDouble()).toInt()

        val board = cell.board
        var i = 1
        while (true) {
            val currentRow = cell.row + deltaRowSign * i
            val currentColumn = cell.column + deltaColumnSign * i

            if (currentRow == destinationCell.row && currentColumn == destinationCell.column) {
                return true
            }

            val currentCell = board.getCell(currentRow, currentColumn)
            if (currentCell.piece != null) {
                return false
            }

            i++
        }
    }

    override fun analyzeAbilityOfEat() {
        canEat = generateSequence(2) { it + 1 }
            .take(6)
            .flatMap { sequenceOf(it to it, it to -it, -it to -it, -it to it) }
            .filter { BoardUtils.isCoordinatesExists(cell.row + it.first, cell.column + it.second) }
            .any { isAbleToEatTo(cell.getNear(it.first, it.second)) }
    }

    override fun eat(destinationCell: Cell) {
        if (!canEat || !isAbleToEatTo(destinationCell)) {
            throw CannotEatException(cell, destinationCell)
        }

        val sacrificedPiece = getSacrificedPiece(destinationCell) ?: throw CannotEatException(cell, destinationCell)
        val sacrificedPieceCell = sacrificedPiece.cell
        cell.piece = null
        sacrificedPieceCell.piece = null
        destinationCell.piece = this
    }

    override fun isAbleToEatTo(destinationCell: Cell): Boolean {
        if (destinationCell.piece != null) {
            return false
        }

        if (isOnOtherDiagonal(destinationCell)) {
            return false
        }

        if (cell.diff(destinationCell) < 2) {
            return false
        }

        val deltaColumn = destinationCell.column - cell.column
        val deltaRow = destinationCell.row - cell.row

        val deltaColumnSign = sign(deltaColumn.toFloat()).toInt()
        val deltaRowSign = sign(deltaRow.toFloat()).toInt()

        var enemyPieceCount = 0
        var i = 1
        while (true) {
            val currentRow = cell.row + deltaRowSign * i
            val currentColumn = cell.column + deltaColumnSign * i
            if (currentColumn == destinationCell.column && currentRow == destinationCell.row) {
                return enemyPieceCount == 1
            }

            val currentCell = cell.board.getCell(currentRow, currentColumn)
            val piece = currentCell.piece
            if (piece == null) {
                i++
                continue
            }

            if (piece.color == color) {
                return false
            }

            enemyPieceCount++
            i++
        }
    }

    private fun isOnOtherDiagonal(destinationCell: Cell): Boolean {
        val deltaColumn = destinationCell.column - cell.column
        val deltaRow = destinationCell.row - cell.row
        val k = deltaColumn.toFloat() / deltaRow
        return abs(k) != 1f
    }

    private fun getSacrificedPiece(destinationCell: Cell): Piece? {
        val destinationRow = destinationCell.row
        val destinationColumn = destinationCell.column
        val sourceRow = cell.row
        val sourceColumn = cell.column

        val signRow = sign((destinationCell.row - cell.row).toFloat()).toInt()
        val signCol = sign((destinationCell.column - cell.column).toFloat()).toInt()

        var i = 1
        while (true) {
            val currentRow = sourceRow + signRow * i
            val currentColumn = sourceColumn + signCol * i
            if (currentRow == destinationRow && currentColumn == destinationColumn) {
                return null
            }

            val currentCell = cell.board.getCell(currentRow, currentColumn)
            val piece = currentCell.piece
            if (piece == null) {
                i++
                continue
            }

            return if (piece.color == color) {
                null
            } else {
                piece
            }
        }
    }

    override fun isMan(): Boolean =
        false

    override fun isKing(): Boolean =
        true

    override fun toString(): String =
        if (color == Color.BLACK) {
            "&"
        } else {
            "#"
        }
}
